File: 슈퍼컴퓨팅문제/KSC2016_ProblemSet_n_Solution.tar/KSC2016_ProblemSet_n_Solution/Prob1.tar/Prob1/Prob1/Code/c_parallel/kmeans.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <mpi.h>
#include "kmeans.h"

void assignment_step(const PPOINT kmeans, const PPOINT pt,
		const int istart, const int iend, int *kindex)
{
	double d1, d2;

	for(int i=istart; i<iend; i++) {
		kindex[i] = 0;
		d1 = distance(pt[i], kmeans[0]);
		for(int j=1; j<N_K; j++) {
			d2 = distance(pt[i], kmeans[j]);
			if( d1 > d2 ) {
				d1 = d2;
				kindex[i] = j;
			}
		}
	}
}

void update_step(PPOINT kmeans, const PPOINT pt,
		const int istart, const int iend, const int *kindex)
{
	int i, idx;
	int num_pt[N_K], local_num_pt[N_K];
	POINT local_kmeans[N_K];

	for(i=0; i<N_K; i++) {
		kmeans[i].x = kmeans[i].y = 0.0;
		local_kmeans[i].x = local_kmeans[i].y = 0.0;
		num_pt[i] = 0;
		local_num_pt[i] = 0;
	}
	for(i=istart; i<iend; i++) {
		idx = kindex[i];
		local_kmeans[idx].x += pt[i].x;
		local_kmeans[idx].y += pt[i].y;
		local_num_pt[idx]++;
	}

	MPI_Allreduce(local_kmeans, kmeans, N_K*2, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD);
	MPI_Allreduce(local_num_pt, num_pt, N_K, MPI_INT, MPI_SUM, MPI_COMM_WORLD);

	for(i=0; i<N_K; i++) {
		kmeans[i].x /= num_pt[i];
		kmeans[i].y /= num_pt[i];
	}
}

File: 슈퍼컴퓨팅문제/KSC2016_ProblemSet_n_Solution.tar/KSC2016_ProblemSet_n_Solution/Prob1.tar/Prob1/Prob1/Code/f_parallel/kmeans.f90
module kmeans_mod
    use util_mod
    implicit none

contains

subroutine assignment_step(kmeans, pt, istart, iend, kindex)
    implicit none
    integer, intent(in) :: istart, iend
    type(POINT), intent(in) :: kmeans(N_K), pt(N_PT)
    integer, intent(out) :: kindex(N_PT)
    integer :: i,j
    real(8) :: d1, d2

    do i=istart, iend
        kindex(i) = 1
        d1 = distance(pt(i), kmeans(1))
        do j=2,N_K
            d2 = distance(pt(i), kmeans(j))
            if( d1 > d2 ) then
                d1 = d2
                kindex(i) = j
            end if
        end do
    end do
end subroutine

subroutine update_step(kmeans, pt, istart, iend, kindex)
    implicit none
	include 'mpif.h'
    integer, intent(in) :: istart, iend, kindex(N_PT)
    type(POINT), intent(in) :: pt(N_PT)
    type(POINT), intent(out) :: kmeans(N_K)
    type(POINT)  :: local_kmeans(N_K)
    integer :: i, idx, num_pt(N_K), local_num_pt(N_K), ierr

    kmeans%x = 0.0
    kmeans%y = 0.0
    local_kmeans%x = 0.0
    local_kmeans%y = 0.0
    num_pt = 0
    local_num_pt = 0

    do i=istart, iend
        idx = kindex(i)
        local_kmeans(idx)%x = local_kmeans(idx)%x + pt(i)%x
        local_kmeans(idx)%y = local_kmeans(idx)%y + pt(i)%y
        local_num_pt(idx) = local_num_pt(idx) + 1
    end do
	
	call MPI_Allreduce(local_kmeans, kmeans, N_K*2, MPI_DOUBLE, MPI_SUM, MPI_COMM_WORLD, ierr)
	call MPI_Allreduce(local_num_pt, num_pt, N_K, MPI_INT, MPI_SUM, MPI_COMM_WORLD, ierr)

    kmeans%x = kmeans%x / num_pt
    kmeans%y = kmeans%y / num_pt
end subroutine

end module
